File: nodejs/groupsandtags/src/namedtags.js
const {CacheManager,CacheItem,NamedTagsDictionary,QueryCommand,JsonDataType} = require('ncache-client');
let cache = null;
const config = require('./app.config.json');
async function Run() {
    // Initialize cache 
    await InitializeCache();
    // Creating named tag dictionary. 
    const namedTagDict = new NamedTagsDictionary();
    namedTagDict.add("Category", "Beverages");
    namedTagDict.add("ProductName", "Coke");

    // Add Items in cache with named tags
    await AddItems(namedTagDict);

    // Fetch Items from the cache
    await GetItems();

}

async function InitializeCache() {
    // Get Cache Name
    const cacheName = config.cacheProperties.cacheName
    if (!cacheName) {
        console.log("The CacheID cannot be null or empty.");
        return;
    }
    // Initialize an instance of the cache to begin performing operations:
    cache = await CacheManager.getCache(cacheName);
    await cache.clear();
    // Print output on console
    console.log(`Cache ${cacheName} is initialized.`);
            
}
async function AddItems(namedTagDict) {
    await Promise.all([
            AddNamedTagDataToCache(14, "Andrew Ng","USA",namedTagDict),
            AddNamedTagDataToCache(15, "Jhon Getsby", "France", namedTagDict)
    ]);
}
async function GetItems() {
    const query = "SELECT $Value$ FROM Alachisoft.NCache.Sample.Data.Order WHERE Category = ? AND ProductName = ?";
    const queryCommand = new QueryCommand(query);
    let parametersMap = new Map();
    parametersMap.set("Category", "Beverages");
    parametersMap.set("ProductName", "Coke");
    queryCommand.setParameters( parametersMap);
    const searchService = await cache.getSearchService();
    const cacheReader = await searchService.executeReader(queryCommand);
    let counter = 0 ;
    if (cacheReader.getFieldCount() > 0) {
        while (await cacheReader.read()) {
            const order = cacheReader.getValue(1,JsonDataType.Object);
            console.log(order);
            counter++;
        }
    }
}
async function AddNamedTagDataToCache(productId, shipName, shipCountry ,namedTagDict) {
    const order = { OrderId : productId, ShipName : shipName, ShipCountry : shipCountry};

    // create object key
    const key = "order:" + order.OrderId;

    // create CacheItem with your desired object
    const  item = new CacheItem(order,"Alachisoft.NCache.Sample.Data.Order");

    // assign group to CacheItem object
    item.setNamedTags(namedTagDict)

    // add CacheItem object to cache
    await cache.add(key, item);
    console.log("items added in cache")
}
module.exports = {
    Run
}

File: python/groups_and_tags/named_tags.py
from ncache.client.CacheManager import CacheManager
from ncache.client.CacheItem import CacheItem
from ncache.runtime.caching.NamedTagsDictionary import NamedTagsDictionary
from ncache.client.QueryCommand import QueryCommand
from SampleData.Product import Product


def initialize_cache():
    cache_name = "demoCache"
    cache = CacheManager.get_cache(cache_name)
    print("\nCache initialized successfully.")
    return cache


def add_items(cache, named_tags):
    product_1 = Product(id="1001", name="Butter", unit_price=50, units_available=20)
    product_2 = Product(id="1002", name="Cheese", unit_price=30, units_available=30)
    products = [product_1, product_2]
    for product in products:
        key = product.id
        item = CacheItem(product)
        item.set_named_tags(named_tags)
        cache.add(key, item)
    print("\nItems are added in cache.")


def get_items(cache):
    query = "SELECT * FROM SampleData.Product.Product WHERE Category = ?"
    query_command = QueryCommand(query)
    parameters = {"Category": "Dairy"}
    query_command.set_parameters(parameters)
    cache_reader = cache.get_search_service().execute_reader(query_command, True, 512)
    if not cache_reader.get_is_closed():
        print("\nThe values fetched are:")
        while cache_reader.read():
            id = cache_reader.get_value(str, None, "id")
            name = cache_reader.get_value(str, None, "name")
            unit_price = cache_reader.get_value(int, None, "unit_price")
            units_available = cache_reader.get_value(int, None, "units_available")
            print(f"\nID: {id}\nName: {name}\nUnit Price: {unit_price}\nUnits Available: {units_available}")


def dispose_cache(cache):
    cache.close()
    print("\nCache closed.")


# Initialize instance of the cache to perform operations
cache = initialize_cache()

# Creating named tags dictionary
named_tags = NamedTagsDictionary()
named_tags.add("Category", "Dairy")

# Add items in cache
add_items(cache, named_tags)

# Get items using named tags
get_items(cache)

# Dispose the cache once done
dispose_cache(cache)
